aspnetcore/blazor/components.md

```diff
@@ -923,3 +923,93 @@ In the following example, the loop in the `CreateComponent` method generates thr
+{
+    builder.AddContent(0, "First");
+}
+
+builder.AddContent(1, "Second");
+```
+
+When this code executes for the first time, if `someFlag` is `true`, the builder receives:
+
+| Sequence | Type      | Data   |
+| :------: | --------- | :----: |
+| 0        | Text node | First  |
+| 1        | Text node | Second |
+
+Now imagine that `someFlag` becomes `false`, and we render again. This time, the builder receives:
+
+| Sequence | Type       | Data   |
+| :------: | ---------- | :----: |
+| 1        | Text node  | Second |
+
+When the runtime performs a diff, it sees that the item at sequence `0` was removed, so it generates the following trivial *edit script*:
+
+* Remove the first text node.
+
+#### What goes wrong if you generate sequence numbers programmatically
+
+Imagine instead that you wrote the following rendertree builder logic:
+
+```csharp
+var seq = 0;
+
+if (someFlag)
+{
+    builder.AddContent(seq++, "First");
+}
+
+builder.AddContent(seq++, "Second");
+```
+
+Now the first output would be:
+
+| Sequence | Type      | Data   |
+| :------: | --------- | :--- : |
+| 0        | Text node | First  |
+| 1        | Text node | Second |
+
+This outcome is identical to the prior case, so no negative issues exist. On the second render when `someFlag` is `false`, the output is:
+
+| Sequence | Type      | Data   |
+| :------: | --------- | ------ |
+| 0        | Text node | Second |
+
+This time, the diff algorithm sees that *two* changes have occurred, and the algorithm generates the following edit script:
+
+* Change the value of the first text node to `Second`.
+* Remove the second text node.
+
+Generating the sequence numbers has lost all the useful information about where the `if/else` branches and loops were present in the original code. This results in a diff **twice as long** as before.
+
+This is a trivial example. In more realistic cases with complex and deeply nested structures, and especially with loops, the performance cost is more severe. Instead of immediately identifying which loop blocks or branches have been inserted or removed, the diff algorithm has to recurse deeply into the render trees and usually build far longer edit scripts because it is misinformed about how the old and new structures relate to each other.
+
+#### Guidance and conclusions
+
+* App performance suffers if sequence numbers are generated dynamically.
+* The framework can't create its own sequence numbers automatically at runtime because the necessary information doesn't exist unless it's captured at compile time.
+* Don't write long blocks of manually-implemented `RenderTreeBuilder` logic. Prefer `.razor` files and allow the compiler deal sequence numbers. If you're unable to avoid manual `RenderTreeBuilder` logic, split long blocks of code into smaller pieces wrapped in `OpenRegion`/`CloseRegion` calls. Each region has its own separate space of sequence numbers, so you can restart from zero (or any other arbitrary number) inside each region.
```

Comment:
```suggestion
* Don't write long blocks of manually-implemented `RenderTreeBuilder` logic. Prefer `.razor` files and allow the compiler to deal with the sequence numbers. If you're unable to avoid manual `RenderTreeBuilder` logic, split long blocks of code into smaller pieces wrapped in `OpenRegion`/`CloseRegion` calls. Each region has its own separate space of sequence numbers, so you can restart from zero (or any other arbitrary number) inside each region.
```